stl/inc/xstring

```diff
@@ -3165,21 +3159,24 @@ public:
                 auto&& _Right_alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Right_al);
                 _Container_proxy_ptr<_Alty> _New_proxy(_Right_alproxy, _Leave_proxy_unbound{}); // throws
 
-                if (_Right._Mypair._Myval2._Large_string_engaged()) {
-                    const auto _New_size     = _Right._Mypair._Myval2._Mysize;
-                    const auto _New_capacity = _Calculate_growth(_New_size, 0, _Right.max_size());
+                const auto _Right_size        = _Right._Mypair._Myval2._Mysize;
+                const _Elem* const _Right_ptr = _Right._Mypair._Myval2._Myptr();
+                if (_Entails_large_string(_Right_size)) {
+                    const auto _New_capacity =
+                        _Calculate_growth(_Right_size, _SMALL_STRING_CAPACITY, _Right.max_size());
                     auto _Right_al_non_const = _Right_al;
-                    const auto _New_ptr      = _Right_al_non_const.allocate(_New_capacity + 1); // throws
-
-                    _Start_element_lifetimes(_Unfancy(_New_ptr), _New_capacity + 1);
+                    const auto _New_ptr      = _Allocate_for_capacity(_Right_al_non_const, _New_capacity); // throws
 
-                    _Traits::copy(_Unfancy(_New_ptr), _Unfancy(_Right._Mypair._Myval2._Bx._Ptr), _New_size + 1);
+                    _Traits::copy(_Unfancy(_New_ptr), _Right_ptr, _Right_size + 1);
                     _Tidy_deallocate();
                     _Mypair._Myval2._Bx._Ptr = _New_ptr;
-                    _Mypair._Myval2._Mysize  = _New_size;
+                    _Mypair._Myval2._Mysize  = _Right_size;
                     _Mypair._Myval2._Myres   = _New_capacity;
                 } else {
-                    _Copy_assign_val_from_small(_Right);
+                    _Tidy_deallocate();
+                    _Traits::copy(_Mypair._Myval2._Bx._Buf, _Right_ptr, _Right_size + 1);
+                    _Mypair._Myval2._Mysize = _Right_size;
+                    _Mypair._Myval2._Myres  = _SMALL_STRING_CAPACITY;

```

Comment:
I think dropping `_Memcpy_val_from` optimization (was in `_Copy_assign_val_from_small`) is ok here, as this branch looks rarely used.
Otherwise we have to deal with three cases:
right is in small mode(res) and in non-constexpr context (and `!defined(_INSERT_STRING_ANNOTATION)`) -> can do `_Memcpy_val_from`
right is small(size) -> copy str without allocation
right is large(size) -> allocate and copy str

Reply:
Yeah, this is in the highly obscure POCCA codepath for non-equal allocators. Correctness is fine, we don't need to micro-optimize this case.